coastguard/coastguard.md

```diff
@@ -0,0 +1,138 @@
+
+We will have 4 main flows:
+  1.	a clusters joins the cluster set 
+  2.	a deployment change happens on any cluster  
+  3.	a user adds/changes new multi-network policy is added
+  4.	a cluster is removed from the set 
+
+Flows 2-3 have no special additions beyond executing the reconciliation loop.
+For flow 1 - when a cluster first joins the cluster set, and the coastguard controller runs for the first time, the controller will need to map all existing namespaces and pods on the k8s cluster. It will do so by running the API equivalent of `kubectl get namespaces` and `kubectl get pods`.
+
+For flow 4 - once a cluster leaves the cluster set, the coastguard controller on the leaving cluster should delete all it’s entries from etcd. 
+
+## Component/Service Coast guard controller
+
+1. ETCD structure
+In general, etcd is a key value store. It has the ability to query based on key prefixes.
+So, for each pod, we will have the following entries, where `$clustername` for example is a placeholder for clustername.
+1.	`$clustername_$namespace_$podname`: ip address
+2.	`label_$label_$clustername_$namespace_$podname`: ip address 
+
+We will need the following queries:
+1.	get all pod ip, in cluster X with namespace Y
+  query: etcdctl get --prefix `$clustername_$namespace`
+2.	get all pod ips  in cluster X with namespace Y with label Z
+  query:  etcdctl get --prefix `label_$labelZ_$clusternameX_$namespaceY`
+
+3.	get all pod ips  in cluster with label Z
+  query:  etcdctl get --prefix `label_$labelZ_$clusternameX`
+  Same idea as 2.
+
+#note: ETCD is holding all the sensitive data in our network, the ETCD security is out of scoop in this project.
+2.	CRD schema
+our multi-cluster network policy CRD:
+```
+apiVersion: networking.submariner.io/v1
+kind: MultiClusterNetworkPolicy
+metadata:
+  name: test-network-policy
+  namespace: default
+spec:
+  clusterSelector: cluster1
+  podSelector:
+    matchLabels:
+      role: db
+
+  policyTypes:
+    ingress:
+    - from:
+      - ipBlock:
+          cidr: 172.17.0.0/16
+          except:
+          - 172.17.1.0/24
+      - namespaceSelector:
+          matchLabels:
+            key: value
+      - podSelector:
+          matchLabels:
+            key: value
+      - clusterSelector: value
+        
+      ports:
+      - protocol: TCP
+        port: 6379
+    egress:
+    - to:
+      - ipBlock:
+          cidr: 172.17.1.0/16
+          except:
+          - 172.17.2.0/24
+      - namespaceSelector:
+          matchLabels:
+            key: value
+      - podSelector:
+          matchLabels:
+            key: value
+      - clusterSelector: value
+```
+
+## Open Questions
+
+1. should we forgo the use of clusterSelector and use just namespaces based on the principal of [namespace sameness](https://github.com/kubernetes/community/blob/master/sig-multicluster/namespace-sameness-position-statement.md)
```

Comment:
We could initially forego cluster filtering and just allow selecting based on namespace/pod/service or whatever we put as first priority, and later on add capability to select based on cluster.